File: app/src/main/java/ru/sukharev/myrecipes/recipelist/RecipeListPresenter.java
package ru.sukharev.myrecipes.recipelist;

import android.content.Context;

import java.util.List;

import ru.sukharev.myrecipes.model.IModel;
import ru.sukharev.myrecipes.model.RecipeRepository;
import ru.sukharev.myrecipes.pojo.Recipe;

/**
 * Presenter for view which shows list of all recipes
 */

public class RecipeListPresenter implements RecipeListContract.Presenter{

    private static RecipeListPresenter instance;

    private RecipeListContract.View view;

    private IModel recipeModel;

    private RecipeListPresenter(Context context) {
        recipeModel = RecipeRepository.getInstance(context.getApplicationContext());
    }

    static RecipeListPresenter init(Context context, RecipeListContract.View view){
        if (instance == null) {
            instance = new RecipeListPresenter(context);
        }
        instance.setView(view);
        return instance;
    }


    private void setView(RecipeListContract.View view){
        if (!view.equals(this.view)) {
            instance.view = view;
            instance.view.setPresenter(this);
        }
    }

    @Override
    public void listItemClicked() {

    }

    @Override
    public void start() {
        recipeModel.getRecipes(new IModel.GetRecipesCallback() {
            @Override
            public void onResult(List<Recipe> recipes) {
                view.showList(recipes);
            }
        });
    }


}

File: app/src/main/java/ru/sukharev/myrecipes/recipelist/RecipeListPresenter.kt
package ru.sukharev.myrecipes.recipelist

import android.content.Context

import ru.sukharev.myrecipes.model.IModel
import ru.sukharev.myrecipes.model.RecipeRepository
import ru.sukharev.myrecipes.pojo.Recipe

/**
 * Presenter for view which shows list of all recipes
 */

class RecipeListPresenter private constructor(context: Context) : RecipeListContract.Presenter {

    private var view: RecipeListContract.View? = null

    private val recipeModel: IModel

    init {
        recipeModel = RecipeRepository.getInstance(context.applicationContext)
    }


    private fun setView(view: RecipeListContract.View) {
        if (view != this.view) {
            instance.view = view
            instance.view!!.setPresenter(this)
        }
    }

    override fun listItemClicked() {

    }

    override fun start() {
        recipeModel.getRecipes(object : IModel.GetRecipesCallback {
            override fun onResult(recipes: List<Recipe>) {
                view!!.showList(recipes)
            }
        })
    }

    companion object {

        private lateinit var instance: RecipeListPresenter

        internal fun init(context: Context, view: RecipeListContract.View): RecipeListPresenter {
            if (instance == null) {
                instance = RecipeListPresenter(context)
            }
            instance.setView(view)
            return instance
        }
    }


}
